common/src/main/java/io/pravega/common/util/AsyncIterator.java

```diff
@@ -56,6 +58,29 @@
                 }, executor);
     }
 
+    /**
+     * Processes the remaining elements in the AsyncIterator until the specified {@link Predicate} returns true.
+     *
+     * @param until A Predicate that will be used to the decide the number of elements on which the Consumer will be invoked.
+     * @param consumer A Consumer that will be invoked for each remaining element. The consumer will be invoked using the
+     *                 given Executor, but any new invocation will wait for the previous invocation to complete.
+     * @param executor An Executor to run async tasks on.
+     * @return A CompletableFuture that, when completed, will indicate that the processing is complete.
+     */
+    default CompletableFuture<Void> forEachRemaining(Predicate<? super T> until, Consumer<? super T> consumer, Executor executor) {
```

Comment:
My suggestion was simpler. You don't need both a Predicate and a Consumer. Just add this method to this interface, which will work without the need for an executor. It will execute the Predicate on the original threadpool, which should be fine since it's not supposed to do much computation or blocking IO anyway:
```
default CompletableFuture<Void> collectRemaining(Predicate<? super T> collector) {
        return getNext().thenCompose(e -> {
            boolean canContinue = e != null && collector.test(e);
            if (canContinue) {
                return forEachRemaining(consumer);
            } else {
                return CompletableFuture.completedFuture(null);
            }
        });
    }
```
https://github.com/pravega/pravega/pull/3337#discussion_r254382968

Reply:
I have updated the method to remove the need for an executor. This will ensure the subsequent `getNext()` is invoked on the same thread pool and all  the entries are collected.

Reply:
refactored the code and removed redundant consumer.